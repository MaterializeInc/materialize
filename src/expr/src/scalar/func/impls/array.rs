// Copyright Materialize, Inc. and contributors. All rights reserved.
//
// Use of this software is governed by the Business Source License
// included in the LICENSE file.
//
// As of the Change Date specified in that file, in accordance with
// the Business Source License, use of this software will be governed
// by the Apache License, Version 2.0.

use std::fmt;

use proptest_derive::Arbitrary;
use serde::{Deserialize, Serialize};

use mz_lowertest::MzReflect;
use mz_repr::{ColumnType, Datum, RowArena, ScalarType};

use crate::scalar::func::{stringify_datum, LazyUnaryFunc};
use crate::{EvalError, MirScalarExpr};

#[derive(
    Arbitrary, Ord, PartialOrd, Clone, Debug, Eq, PartialEq, Serialize, Deserialize, Hash, MzReflect,
)]
pub struct CastArrayToListOneDim;

impl LazyUnaryFunc for CastArrayToListOneDim {
    fn eval<'a>(
        &'a self,
        datums: &[Datum<'a>],
        temp_storage: &'a RowArena,
        a: &'a MirScalarExpr,
    ) -> Result<Datum<'a>, EvalError> {
        let a = a.eval(datums, temp_storage)?;
        if a.is_null() {
            return Ok(Datum::Null);
        }

        let arr = a.unwrap_array();
        let ndims = arr.dims().ndims();
        if ndims > 1 {
            return Err(EvalError::Unsupported {
                feature: format!(
                    "casting multi-dimensional array to list; got array with {} dimensions",
                    ndims
                ),
                issue_no: None,
            });
        }

        Ok(Datum::List(arr.elements()))
    }

    /// The output ColumnType of this function
    fn output_type(&self, input_type: ColumnType) -> ColumnType {
        ScalarType::List {
            element_type: Box::new(input_type.scalar_type.unwrap_array_element_type().clone()),
            custom_id: None,
        }
        .nullable(true)
    }

    /// Whether this function will produce NULL on NULL input
    fn propagates_nulls(&self) -> bool {
        true
    }

    /// Whether this function will produce NULL on non-NULL input
    fn introduces_nulls(&self) -> bool {
        false
    }

    /// Whether this function preserves uniqueness
    fn preserves_uniqueness(&self) -> bool {
        true
    }

    fn inverse(&self) -> Option<crate::UnaryFunc> {
        None
    }
}

impl fmt::Display for CastArrayToListOneDim {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        f.write_str("arraytolist")
    }
}

#[derive(
    Arbitrary, Ord, PartialOrd, Clone, Debug, Eq, PartialEq, Serialize, Deserialize, Hash, MzReflect,
)]
pub struct CastArrayToString {
    pub ty: ScalarType,
}

impl LazyUnaryFunc for CastArrayToString {
    fn eval<'a>(
        &'a self,
        datums: &[Datum<'a>],
        temp_storage: &'a RowArena,
        a: &'a MirScalarExpr,
    ) -> Result<Datum<'a>, EvalError> {
        let a = a.eval(datums, temp_storage)?;
        if a.is_null() {
            return Ok(Datum::Null);
        }
        let mut buf = String::new();
        stringify_datum(&mut buf, a, &self.ty)?;
        Ok(Datum::String(temp_storage.push_string(buf)))
    }

    fn output_type(&self, input_type: ColumnType) -> ColumnType {
        ScalarType::String.nullable(input_type.nullable)
    }

    fn propagates_nulls(&self) -> bool {
        true
    }

    fn introduces_nulls(&self) -> bool {
        false
    }

    fn preserves_uniqueness(&self) -> bool {
        true
    }

    fn inverse(&self) -> Option<crate::UnaryFunc> {
        // TODO? If we moved typeconv into `expr` we could determine the right
        // inverse of this.
        None
    }
}

impl fmt::Display for CastArrayToString {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        f.write_str("arraytostr")
    }
}
