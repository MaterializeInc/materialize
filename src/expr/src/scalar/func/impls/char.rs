// Copyright Materialize, Inc. and contributors. All rights reserved.
//
// Use of this software is governed by the Business Source License
// included in the LICENSE file.
//
// As of the Change Date specified in that file, in accordance with
// the Business Source License, use of this software will be governed
// by the Apache License, Version 2.0.

use std::fmt;

use proptest_derive::Arbitrary;
use serde::{Deserialize, Serialize};

use mz_lowertest::MzReflect;
use mz_repr::adt::char::{format_str_pad, Char, CharLength};
use mz_repr::{ColumnType, ScalarType};

use crate::scalar::func::EagerUnaryFunc;

/// All Char data is stored in Datum::String with its blank padding removed
/// (i.e. trimmed), so this function provides a means of restoring any
/// removed padding.
#[derive(
    Arbitrary, Ord, PartialOrd, Clone, Debug, Eq, PartialEq, Serialize, Deserialize, Hash, MzReflect,
)]
pub struct PadChar {
    pub length: Option<CharLength>,
}

impl<'a> EagerUnaryFunc<'a> for PadChar {
    type Input = &'a str;
    type Output = Char<String>;

    fn call(&self, a: &'a str) -> Char<String> {
        Char(format_str_pad(a, self.length))
    }

    fn output_type(&self, input: ColumnType) -> ColumnType {
        ScalarType::Char {
            length: self.length,
        }
        .nullable(input.nullable)
    }
}

impl fmt::Display for PadChar {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        f.write_str("padchar")
    }
}

// This function simply allows the expression of changing a's type from varchar to string
sqlfunc!(
    #[sqlname = "char_to_text"]
    #[preserves_uniqueness = true]
    fn cast_char_to_string<'a>(a: Char<&'a str>) -> &'a str {
        a.0
    }
);
