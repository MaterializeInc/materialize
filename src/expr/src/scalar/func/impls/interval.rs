// Copyright Materialize, Inc. and contributors. All rights reserved.
//
// Use of this software is governed by the Business Source License
// included in the LICENSE file.
//
// As of the Change Date specified in that file, in accordance with
// the Business Source License, use of this software will be governed
// by the Apache License, Version 2.0.

use chrono::NaiveTime;
use num::traits::CheckedNeg;

use crate::EvalError;
use mz_repr::adt::interval::Interval;
use mz_repr::strconv;

sqlfunc!(
    #[sqlname = "interval_to_text"]
    #[preserves_uniqueness = true]
    #[inverse = to_unary!(super::CastStringToInterval)]
    fn cast_interval_to_string(a: Interval) -> String {
        let mut buf = String::new();
        strconv::format_interval(&mut buf, a);
        buf
    }
);

sqlfunc!(
    #[sqlname = "interval_to_time"]
    #[preserves_uniqueness = false]
    #[inverse = to_unary!(super::CastTimeToInterval)]
    fn cast_interval_to_time(mut i: Interval) -> NaiveTime {
        // Negative durations have their HH::MM::SS.NS values subtracted from 1 day.
        if i.is_negative() {
            i = Interval::new(0, 0, 86_400_000_000)
                .checked_add(&i.as_time_interval())
                .unwrap();
        }

        // TODO(benesch): remove potentially dangerous usage of `as`.
        #[allow(clippy::as_conversions)]
        NaiveTime::from_hms_nano_opt(
            i.hours() as u32,
            i.minutes() as u32,
            i.seconds::<f64>() as u32,
            i.nanoseconds() as u32,
        )
        .unwrap()
    }
);

sqlfunc!(
    #[sqlname = "-"]
    #[preserves_uniqueness = true]
    #[inverse = to_unary!(super::NegInterval)]
    fn neg_interval(i: Interval) -> Result<Interval, EvalError> {
        i.checked_neg().ok_or(EvalError::IntervalOutOfRange)
    }
);

sqlfunc!(
    #[sqlname = "justify_days"]
    fn justify_days(i: Interval) -> Result<Interval, EvalError> {
        i.justify_days().map_err(|_| EvalError::IntervalOutOfRange)
    }
);

sqlfunc!(
    #[sqlname = "justify_hours"]
    fn justify_hours(i: Interval) -> Result<Interval, EvalError> {
        i.justify_hours().map_err(|_| EvalError::IntervalOutOfRange)
    }
);

sqlfunc!(
    #[sqlname = "justify_interval"]
    fn justify_interval(i: Interval) -> Result<Interval, EvalError> {
        i.justify_interval()
            .map_err(|_| EvalError::IntervalOutOfRange)
    }
);
