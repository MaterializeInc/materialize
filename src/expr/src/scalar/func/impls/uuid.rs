// Copyright Materialize, Inc. and contributors. All rights reserved.
//
// Use of this software is governed by the Business Source License
// included in the LICENSE file.
//
// As of the Change Date specified in that file, in accordance with
// the Business Source License, use of this software will be governed
// by the Apache License, Version 2.0.

use uuid::Uuid;

use mz_repr::strconv;

sqlfunc!(
    #[sqlname = "uuid_to_text"]
    #[preserves_uniqueness = true]
    #[inverse = to_unary!(super::CastStringToUuid)]
    fn cast_uuid_to_string(u: Uuid) -> String {
        let mut buf = String::with_capacity(36);
        strconv::format_uuid(&mut buf, u);
        buf
    }
);
