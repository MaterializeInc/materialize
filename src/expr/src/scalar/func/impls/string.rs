// Copyright Materialize, Inc. and contributors. All rights reserved.
//
// Use of this software is governed by the Business Source License
// included in the LICENSE file.
//
// As of the Change Date specified in that file, in accordance with
// the Business Source License, use of this software will be governed
// by the Apache License, Version 2.0.

use std::borrow::Cow;
use std::fmt;

use chrono::{DateTime, NaiveDateTime, NaiveTime, Utc};
use mz_repr::adt::date::Date;
use mz_repr::adt::timestamp::CheckedTimestamp;
use once_cell::sync::Lazy;
use proptest_derive::Arbitrary;
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use mz_lowertest::MzReflect;
use mz_ore::cast::CastFrom;
use mz_ore::result::ResultExt;
use mz_ore::str::StrExt;
use mz_repr::adt::array::ArrayDimension;
use mz_repr::adt::char::{format_str_trim, Char};
use mz_repr::adt::interval::Interval;
use mz_repr::adt::jsonb::Jsonb;
use mz_repr::adt::numeric::{self, Numeric, NumericMaxScale};
use mz_repr::adt::regex::Regex;
use mz_repr::adt::system::{Oid, PgLegacyChar};
use mz_repr::adt::varchar::{VarChar, VarCharMaxLength};
use mz_repr::{strconv, ColumnType, Datum, Row, RowArena, ScalarType};

use crate::scalar::func::{array_create_scalar, EagerUnaryFunc, LazyUnaryFunc};
use crate::{like_pattern, EvalError, MirScalarExpr, UnaryFunc};

sqlfunc!(
    #[sqlname = "text_to_boolean"]
    fn cast_string_to_bool<'a>(a: &'a str) -> Result<bool, EvalError> {
        strconv::parse_bool(a).err_into()
    }
);

sqlfunc!(
    #[sqlname = "text_to_\"char\""]
    #[preserves_uniqueness = true]
    fn cast_string_to_pg_legacy_char<'a>(a: &'a str) -> PgLegacyChar {
        PgLegacyChar(a.as_bytes().get(0).copied().unwrap_or(0))
    }
);

sqlfunc!(
    #[sqlname = "text_to_bytea"]
    #[preserves_uniqueness = true]
    fn cast_string_to_bytes<'a>(a: &'a str) -> Result<Vec<u8>, EvalError> {
        strconv::parse_bytes(a).err_into()
    }
);

sqlfunc!(
    #[sqlname = "text_to_smallint"]
    fn cast_string_to_int16<'a>(a: &'a str) -> Result<i16, EvalError> {
        strconv::parse_int16(a).err_into()
    }
);

sqlfunc!(
    #[sqlname = "text_to_integer"]
    fn cast_string_to_int32<'a>(a: &'a str) -> Result<i32, EvalError> {
        strconv::parse_int32(a).err_into()
    }
);

sqlfunc!(
    #[sqlname = "text_to_bigint"]
    fn cast_string_to_int64<'a>(a: &'a str) -> Result<i64, EvalError> {
        strconv::parse_int64(a).err_into()
    }
);

sqlfunc!(
    #[sqlname = "text_to_real"]
    fn cast_string_to_float32<'a>(a: &'a str) -> Result<f32, EvalError> {
        strconv::parse_float32(a).err_into()
    }
);

sqlfunc!(
    #[sqlname = "text_to_double"]
    fn cast_string_to_float64<'a>(a: &'a str) -> Result<f64, EvalError> {
        strconv::parse_float64(a).err_into()
    }
);

sqlfunc!(
    #[sqlname = "text_to_oid"]
    fn cast_string_to_oid<'a>(a: &'a str) -> Result<Oid, EvalError> {
        Ok(Oid(strconv::parse_oid(a)?))
    }
);

sqlfunc!(
    #[sqlname = "text_to_uint2"]
    fn cast_string_to_uint16(a: &'a str) -> Result<u16, EvalError> {
        strconv::parse_uint16(a).err_into()
    }
);

sqlfunc!(
    #[sqlname = "text_to_uint4"]
    fn cast_string_to_uint32(a: &'a str) -> Result<u32, EvalError> {
        strconv::parse_uint32(a).err_into()
    }
);

sqlfunc!(
    #[sqlname = "text_to_uint8"]
    fn cast_string_to_uint64(a: &'a str) -> Result<u64, EvalError> {
        strconv::parse_uint64(a).err_into()
    }
);

#[derive(
    Arbitrary, Ord, PartialOrd, Clone, Debug, Eq, PartialEq, Serialize, Deserialize, Hash, MzReflect,
)]
pub struct CastStringToNumeric(pub Option<NumericMaxScale>);

impl<'a> EagerUnaryFunc<'a> for CastStringToNumeric {
    type Input = &'a str;
    type Output = Result<Numeric, EvalError>;

    fn call(&self, a: &'a str) -> Result<Numeric, EvalError> {
        let mut d = strconv::parse_numeric(a)?;
        if let Some(scale) = self.0 {
            if numeric::rescale(&mut d.0, scale.into_u8()).is_err() {
                return Err(EvalError::NumericFieldOverflow);
            }
        }
        Ok(d.into_inner())
    }

    fn output_type(&self, input: ColumnType) -> ColumnType {
        ScalarType::Numeric { max_scale: self.0 }.nullable(input.nullable)
    }
}

impl fmt::Display for CastStringToNumeric {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        f.write_str("text_to_numeric")
    }
}

sqlfunc!(
    #[sqlname = "text_to_date"]
    fn cast_string_to_date<'a>(a: &'a str) -> Result<Date, EvalError> {
        strconv::parse_date(a).err_into()
    }
);

sqlfunc!(
    #[sqlname = "text_to_time"]
    fn cast_string_to_time<'a>(a: &'a str) -> Result<NaiveTime, EvalError> {
        strconv::parse_time(a).err_into()
    }
);

sqlfunc!(
    #[sqlname = "text_to_timestamp"]
    fn cast_string_to_timestamp<'a>(
        a: &'a str,
    ) -> Result<CheckedTimestamp<NaiveDateTime>, EvalError> {
        strconv::parse_timestamp(a).err_into()
    }
);

sqlfunc!(
    #[sqlname = "text_to_timestamp_with_time_zone"]
    fn cast_string_to_timestamp_tz<'a>(
        a: &'a str,
    ) -> Result<CheckedTimestamp<DateTime<Utc>>, EvalError> {
        strconv::parse_timestamptz(a).err_into()
    }
);

sqlfunc!(
    #[sqlname = "text_to_interval"]
    fn cast_string_to_interval<'a>(a: &'a str) -> Result<Interval, EvalError> {
        strconv::parse_interval(a).err_into()
    }
);

sqlfunc!(
    #[sqlname = "text_to_uuid"]
    fn cast_string_to_uuid<'a>(a: &'a str) -> Result<Uuid, EvalError> {
        strconv::parse_uuid(a).err_into()
    }
);

#[derive(Ord, PartialOrd, Clone, Debug, Eq, PartialEq, Serialize, Deserialize, Hash, MzReflect)]
pub struct CastStringToArray {
    // Target array's type.
    pub return_ty: ScalarType,
    // The expression to cast the discovered array elements to the array's
    // element type.
    pub cast_expr: Box<MirScalarExpr>,
}

impl LazyUnaryFunc for CastStringToArray {
    fn eval<'a>(
        &'a self,
        datums: &[Datum<'a>],
        temp_storage: &'a RowArena,
        a: &'a MirScalarExpr,
    ) -> Result<Datum<'a>, EvalError> {
        let a = a.eval(datums, temp_storage)?;
        if a.is_null() {
            return Ok(Datum::Null);
        }
        let datums = strconv::parse_array(
            a.unwrap_str(),
            || Datum::Null,
            |elem_text| {
                let elem_text = match elem_text {
                    Cow::Owned(s) => temp_storage.push_string(s),
                    Cow::Borrowed(s) => s,
                };
                self.cast_expr
                    .eval(&[Datum::String(elem_text)], temp_storage)
            },
        )?;
        array_create_scalar(&datums, temp_storage)
    }

    /// The output ColumnType of this function
    fn output_type(&self, input_type: ColumnType) -> ColumnType {
        self.return_ty.clone().nullable(input_type.nullable)
    }

    /// Whether this function will produce NULL on NULL input
    fn propagates_nulls(&self) -> bool {
        true
    }

    /// Whether this function will produce NULL on non-NULL input
    fn introduces_nulls(&self) -> bool {
        false
    }

    /// Whether this function preserves uniqueness
    fn preserves_uniqueness(&self) -> bool {
        false
    }
}

impl fmt::Display for CastStringToArray {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        f.write_str("strtoarray")
    }
}

#[derive(Ord, PartialOrd, Clone, Debug, Eq, PartialEq, Serialize, Deserialize, Hash, MzReflect)]
pub struct CastStringToList {
    // Target list's type
    pub return_ty: ScalarType,
    // The expression to cast the discovered list elements to the list's
    // element type.
    pub cast_expr: Box<MirScalarExpr>,
}

impl LazyUnaryFunc for CastStringToList {
    fn eval<'a>(
        &'a self,
        datums: &[Datum<'a>],
        temp_storage: &'a RowArena,
        a: &'a MirScalarExpr,
    ) -> Result<Datum<'a>, EvalError> {
        let a = a.eval(datums, temp_storage)?;
        if a.is_null() {
            return Ok(Datum::Null);
        }
        let parsed_datums = strconv::parse_list(
            a.unwrap_str(),
            matches!(
                self.return_ty.unwrap_list_element_type(),
                ScalarType::List { .. }
            ),
            || Datum::Null,
            |elem_text| {
                let elem_text = match elem_text {
                    Cow::Owned(s) => temp_storage.push_string(s),
                    Cow::Borrowed(s) => s,
                };
                self.cast_expr
                    .eval(&[Datum::String(elem_text)], temp_storage)
            },
        )?;

        Ok(temp_storage.make_datum(|packer| packer.push_list(parsed_datums)))
    }

    /// The output ColumnType of this function
    fn output_type(&self, _input_type: ColumnType) -> ColumnType {
        self.return_ty.without_modifiers().nullable(false)
    }

    /// Whether this function will produce NULL on NULL input
    fn propagates_nulls(&self) -> bool {
        true
    }

    /// Whether this function will produce NULL on non-NULL input
    fn introduces_nulls(&self) -> bool {
        false
    }

    /// Whether this function preserves uniqueness
    fn preserves_uniqueness(&self) -> bool {
        false
    }
}

impl fmt::Display for CastStringToList {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        f.write_str("strtolist")
    }
}

#[derive(Ord, PartialOrd, Clone, Debug, Eq, PartialEq, Serialize, Deserialize, Hash, MzReflect)]
pub struct CastStringToMap {
    // Target map's value type
    pub return_ty: ScalarType,
    // The expression used to cast the discovered values to the map's value
    // type.
    pub cast_expr: Box<MirScalarExpr>,
}

impl LazyUnaryFunc for CastStringToMap {
    fn eval<'a>(
        &'a self,
        datums: &[Datum<'a>],
        temp_storage: &'a RowArena,
        a: &'a MirScalarExpr,
    ) -> Result<Datum<'a>, EvalError> {
        let a = a.eval(datums, temp_storage)?;
        if a.is_null() {
            return Ok(Datum::Null);
        }
        let parsed_map = strconv::parse_map(
            a.unwrap_str(),
            matches!(
                self.return_ty.unwrap_map_value_type(),
                ScalarType::Map { .. }
            ),
            |value_text| -> Result<Datum, EvalError> {
                let value_text = match value_text {
                    Cow::Owned(s) => temp_storage.push_string(s),
                    Cow::Borrowed(s) => s,
                };
                self.cast_expr
                    .eval(&[Datum::String(value_text)], temp_storage)
            },
        )?;
        let mut pairs: Vec<(String, Datum)> = parsed_map.into_iter().map(|(k, v)| (k, v)).collect();
        pairs.sort_by(|(k1, _v1), (k2, _v2)| k1.cmp(k2));
        pairs.dedup_by(|(k1, _v1), (k2, _v2)| k1 == k2);
        Ok(temp_storage.make_datum(|packer| {
            packer.push_dict_with(|packer| {
                for (k, v) in pairs {
                    packer.push(Datum::String(&k));
                    packer.push(v);
                }
            })
        }))
    }

    /// The output ColumnType of this function
    fn output_type(&self, input_type: ColumnType) -> ColumnType {
        self.return_ty.clone().nullable(input_type.nullable)
    }

    /// Whether this function will produce NULL on NULL input
    fn propagates_nulls(&self) -> bool {
        true
    }

    /// Whether this function will produce NULL on non-NULL input
    fn introduces_nulls(&self) -> bool {
        false
    }

    /// Whether this function preserves uniqueness
    fn preserves_uniqueness(&self) -> bool {
        false
    }
}

impl fmt::Display for CastStringToMap {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        f.write_str("strtomap")
    }
}

#[derive(
    Arbitrary, Ord, PartialOrd, Clone, Debug, Eq, PartialEq, Serialize, Deserialize, Hash, MzReflect,
)]
pub struct CastStringToChar {
    pub length: Option<mz_repr::adt::char::CharLength>,
    pub fail_on_len: bool,
}

impl<'a> EagerUnaryFunc<'a> for CastStringToChar {
    type Input = &'a str;
    type Output = Result<Char<String>, EvalError>;

    fn call(&self, a: &'a str) -> Result<Char<String>, EvalError> {
        let s = format_str_trim(a, self.length, self.fail_on_len).map_err(|_| {
            assert!(self.fail_on_len);
            EvalError::StringValueTooLong {
                target_type: "character".to_string(),
                length: usize::cast_from(self.length.unwrap().into_u32()),
            }
        })?;

        Ok(Char(s))
    }

    fn output_type(&self, input: ColumnType) -> ColumnType {
        ScalarType::Char {
            length: self.length,
        }
        .nullable(input.nullable)
    }
}

impl fmt::Display for CastStringToChar {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        f.write_str("text_to_char")
    }
}

#[derive(
    Arbitrary, Ord, PartialOrd, Clone, Debug, Eq, PartialEq, Serialize, Deserialize, Hash, MzReflect,
)]
pub struct CastStringToVarChar {
    pub length: Option<VarCharMaxLength>,
    pub fail_on_len: bool,
}

impl<'a> EagerUnaryFunc<'a> for CastStringToVarChar {
    type Input = &'a str;
    type Output = Result<VarChar<String>, EvalError>;

    fn call(&self, a: &'a str) -> Result<VarChar<String>, EvalError> {
        let s =
            mz_repr::adt::varchar::format_str(a, self.length, self.fail_on_len).map_err(|_| {
                assert!(self.fail_on_len);
                EvalError::StringValueTooLong {
                    target_type: "character varying".to_string(),
                    length: usize::cast_from(self.length.unwrap().into_u32()),
                }
            })?;

        Ok(VarChar(s))
    }

    fn output_type(&self, input: ColumnType) -> ColumnType {
        ScalarType::VarChar {
            max_length: self.length,
        }
        .nullable(input.nullable)
    }
}

impl fmt::Display for CastStringToVarChar {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        f.write_str("text_to_varchar")
    }
}

// If we support another vector type, this should likely get hoisted into a
// position akin to array parsing.
static INT2VECTOR_CAST_EXPR: Lazy<MirScalarExpr> = Lazy::new(|| MirScalarExpr::CallUnary {
    func: UnaryFunc::CastStringToInt16(CastStringToInt16),
    expr: Box::new(MirScalarExpr::Column(0)),
});

#[derive(
    Arbitrary, Ord, PartialOrd, Clone, Debug, Eq, PartialEq, Serialize, Deserialize, Hash, MzReflect,
)]
pub struct CastStringToInt2Vector;

impl LazyUnaryFunc for CastStringToInt2Vector {
    fn eval<'a>(
        &'a self,
        datums: &[Datum<'a>],
        temp_storage: &'a RowArena,
        a: &'a MirScalarExpr,
    ) -> Result<Datum<'a>, EvalError> {
        let a = a.eval(datums, temp_storage)?;
        if a.is_null() {
            return Ok(Datum::Null);
        }

        let datums = strconv::parse_legacy_vector(a.unwrap_str(), |elem_text| {
            let elem_text = match elem_text {
                Cow::Owned(s) => temp_storage.push_string(s),
                Cow::Borrowed(s) => s,
            };
            INT2VECTOR_CAST_EXPR.eval(&[Datum::String(elem_text)], temp_storage)
        })?;
        array_create_scalar(&datums, temp_storage)
    }

    /// The output ColumnType of this function
    fn output_type(&self, _input_type: ColumnType) -> ColumnType {
        ScalarType::Int2Vector.nullable(false)
    }

    /// Whether this function will produce NULL on NULL input
    fn propagates_nulls(&self) -> bool {
        true
    }

    /// Whether this function will produce NULL on non-NULL input
    fn introduces_nulls(&self) -> bool {
        false
    }

    /// Whether this function preserves uniqueness
    fn preserves_uniqueness(&self) -> bool {
        false
    }
}

impl fmt::Display for CastStringToInt2Vector {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        f.write_str("strtoint2vector")
    }
}

sqlfunc!(
    #[sqlname = "text_to_jsonb"]
    // TODO(jamii): it would be much more efficient to skip the intermediate repr::jsonb::Jsonb.
    fn cast_string_to_jsonb<'a>(a: &'a str) -> Result<Jsonb, EvalError> {
        Ok(strconv::parse_jsonb(a)?)
    }
);

sqlfunc!(
    #[sqlname = "btrim"]
    fn trim_whitespace<'a>(a: &'a str) -> &'a str {
        a.trim_matches(' ')
    }
);

sqlfunc!(
    #[sqlname = "ltrim"]
    fn trim_leading_whitespace<'a>(a: &'a str) -> &'a str {
        a.trim_start_matches(' ')
    }
);

sqlfunc!(
    #[sqlname = "rtrim"]
    fn trim_trailing_whitespace<'a>(a: &'a str) -> &'a str {
        a.trim_end_matches(' ')
    }
);

sqlfunc!(
    #[sqlname = "ascii"]
    fn ascii<'a>(a: &'a str) -> i32 {
        match a.chars().next() {
            None => 0,
            Some(v) => v as i32,
        }
    }
);

sqlfunc!(
    #[sqlname = "char_length"]
    fn char_length<'a>(a: &'a str) -> Result<i32, EvalError> {
        i32::try_from(a.chars().count()).or(Err(EvalError::Int32OutOfRange))
    }
);

sqlfunc!(
    #[sqlname = "bit_length"]
    fn bit_length_string<'a>(a: &'a str) -> Result<i32, EvalError> {
        i32::try_from(a.as_bytes().len() * 8).or(Err(EvalError::Int32OutOfRange))
    }
);

sqlfunc!(
    #[sqlname = "octet_length"]
    fn byte_length_string<'a>(a: &'a str) -> Result<i32, EvalError> {
        i32::try_from(a.as_bytes().len()).or(Err(EvalError::Int32OutOfRange))
    }
);

sqlfunc!(
    fn upper<'a>(a: &'a str) -> String {
        a.to_uppercase()
    }
);

sqlfunc!(
    fn lower<'a>(a: &'a str) -> String {
        a.to_lowercase()
    }
);

#[derive(Ord, PartialOrd, Clone, Debug, Eq, PartialEq, Serialize, Deserialize, Hash, MzReflect)]
pub struct IsLikeMatch(pub like_pattern::Matcher);

impl<'a> EagerUnaryFunc<'a> for IsLikeMatch {
    type Input = &'a str;
    type Output = bool;

    fn call(&self, haystack: &'a str) -> bool {
        self.0.is_match(haystack)
    }

    fn output_type(&self, input: ColumnType) -> ColumnType {
        ScalarType::Bool.nullable(input.nullable)
    }
}

impl fmt::Display for IsLikeMatch {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{} ~~", self.0.pattern.quoted())
    }
}

#[derive(Ord, PartialOrd, Clone, Debug, Eq, PartialEq, Serialize, Deserialize, Hash, MzReflect)]
pub struct IsRegexpMatch(pub Regex);

impl<'a> EagerUnaryFunc<'a> for IsRegexpMatch {
    type Input = &'a str;
    type Output = bool;

    fn call(&self, haystack: &'a str) -> bool {
        self.0.is_match(haystack)
    }

    fn output_type(&self, input: ColumnType) -> ColumnType {
        ScalarType::Bool.nullable(input.nullable)
    }
}

impl fmt::Display for IsRegexpMatch {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{} ~", self.0.as_str().quoted())
    }
}

#[derive(Ord, PartialOrd, Clone, Debug, Eq, PartialEq, Serialize, Deserialize, Hash, MzReflect)]
pub struct RegexpMatch(pub Regex);

impl LazyUnaryFunc for RegexpMatch {
    fn eval<'a>(
        &'a self,
        datums: &[Datum<'a>],
        temp_storage: &'a RowArena,
        a: &'a MirScalarExpr,
    ) -> Result<Datum<'a>, EvalError> {
        let haystack = a.eval(datums, temp_storage)?;
        if haystack.is_null() {
            return Ok(Datum::Null);
        }
        let mut row = Row::default();
        let mut packer = row.packer();
        if self.0.captures_len() > 1 {
            // The regex contains capture groups, so return an array containing the
            // matched text in each capture group, unless the entire match fails.
            // Individual capture groups may also be null if that group did not
            // participate in the match.
            match self.0.captures(haystack.unwrap_str()) {
                None => packer.push(Datum::Null),
                Some(captures) => packer.push_array(
                    &[ArrayDimension {
                        lower_bound: 1,
                        length: captures.len() - 1,
                    }],
                    // Skip the 0th capture group, which is the whole match.
                    captures.iter().skip(1).map(|mtch| match mtch {
                        None => Datum::Null,
                        Some(mtch) => Datum::String(mtch.as_str()),
                    }),
                )?,
            }
        } else {
            // The regex contains no capture groups, so return a one-element array
            // containing the match, or null if there is no match.
            match self.0.find(haystack.unwrap_str()) {
                None => packer.push(Datum::Null),
                Some(mtch) => packer.push_array(
                    &[ArrayDimension {
                        lower_bound: 1,
                        length: 1,
                    }],
                    std::iter::once(Datum::String(mtch.as_str())),
                )?,
            };
        };
        Ok(temp_storage.push_unary_row(row))
    }

    /// The output ColumnType of this function
    fn output_type(&self, _input_type: ColumnType) -> ColumnType {
        ScalarType::Array(Box::new(ScalarType::String)).nullable(true)
    }

    /// Whether this function will produce NULL on NULL input
    fn propagates_nulls(&self) -> bool {
        true
    }

    /// Whether this function will produce NULL on non-NULL input
    fn introduces_nulls(&self) -> bool {
        // Returns null if the regex did not match
        true
    }

    /// Whether this function preserves uniqueness
    fn preserves_uniqueness(&self) -> bool {
        false
    }
}

impl fmt::Display for RegexpMatch {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "regexp_match[{}]", self.0.as_str())
    }
}

sqlfunc!(
    #[sqlname = "mz_panic"]
    fn panic<'a>(a: &'a str) -> String {
        print!("{}", a);
        panic!("{}", a)
    }
);
