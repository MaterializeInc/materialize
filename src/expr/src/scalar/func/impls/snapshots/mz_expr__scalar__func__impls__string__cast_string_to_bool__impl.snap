---
source: src/expr/src/scalar/func/impls/string.rs
expression: info
---
Info {
    output_type_nullable: ColumnType {
        scalar_type: Bool,
        nullable: true,
    },
    output_type_nonnullable: ColumnType {
        scalar_type: Bool,
        nullable: false,
    },
    preserves_uniqueness: false,
    inverse: Some(
        CastBoolToString(
            CastBoolToString,
        ),
    ),
    is_monotone: false,
    propagates_nulls: true,
    introduces_nulls: false,
    could_error: true,
}
